rocksdb_admin/application_db_backup_manager.cpp

```diff
@@ -185,17 +250,86 @@ bool ApplicationDBBackupManager::backupDBToS3(const std::shared_ptr<ApplicationD
+
+  std::unordered_map<std::string, int64_t> file_to_ts;
+  const string sst_suffix = ".sst";
+
+  if (last_ts >= 0) {
+    LOG(INFO) << "find the last backup on " << last_ts;
+    std::string local_path_last_backup = folly::stringPrintf("%ss3_tmp/%s%ld/", rocksdb_dir_.c_str(), db->db_name().c_str(), last_ts) + "checkpoint/";
+    std::string formatted_s3_last_backup_dir_path = folly::stringPrintf("%s%s/%ld/", 
+      ensure_ends_with_pathsep(FLAGS_s3_incre_backup_prefix).c_str(), db->db_name().c_str(), last_ts);
+
+    boost::system::error_code remove_last_backup_err;
+    boost::system::error_code create_last_backup_err;
+    // if we have not removed the last backup, we do not need to download it
+    bool need_download = !boost::filesystem::exists(local_path_last_backup + backupDesc);
+    boost::filesystem::create_directories(local_path_last_backup, create_last_backup_err);
+    SCOPE_EXIT { boost::filesystem::remove_all(local_path_last_backup, remove_last_backup_err); };
+    if (create_err) {
+      // SetException("Cannot remove/create dir for backup: " + local_path, AdminErrorCode::DB_ADMIN_ERROR, &callback);
+      common::Stats::get()->Incr(kS3BackupFailure);
+      return false;
+    }
+
+    if (need_download) {
+      auto resp = local_s3_util->getObject(formatted_s3_last_backup_dir_path + backupDesc, 
+        local_path_last_backup + backupDesc, FLAGS_s3_direct_io);
+    } else {
+      assert(boost::filesystem::exists(local_path_last_backup + backupDesc));
+    }
+
+    std::string last_backup_desc_contents;
+    common::FileUtil::readFileToString(local_path_last_backup + backupDesc, &last_backup_desc_contents);
+    
+    parseBackupDesc(last_backup_desc_contents, &file_to_ts);
+    if (file_to_ts.empty()) {
+      LOG(INFO) << "The last backup of " << db->db_name() << " on timestamp " << last_ts << " is broken"; 
+    } 
+
+    // Remove all the duplicate files
+    checkpoint_files.erase(std::remove_if(checkpoint_files.begin(), checkpoint_files.end(), [&](string file){
+      return file_to_ts.find(file) != file_to_ts.end() && ends_with(file, sst_suffix);}), checkpoint_files.end());
+
+    LOG(INFO) << "finish parsing last backup";
+  } 
+
+  // Get the min_id and max_id for backup descriptor
+  int64_t sst_id_min = INT32_MAX;
+  int64_t sst_id_max = 0;
+  for (auto& file : checkpoint_files) {
+    if (ends_with(file, sst_suffix)) {
+      std::string file_id = parse_sst_id(file);
+      int64_t file_num = remove_leading_zero(file_id);
+      file_to_ts.emplace(file, ts);
+      sst_id_max = std::max(file_num, sst_id_max);
+      sst_id_min = std::min(file_num, sst_id_min);
+    }
+  }
+
+  // Create the new backup descriptor file
+  std::string backup_desc_contents;
+  makeUpBackupDescString(file_to_ts, sst_id_min, sst_id_max, backup_desc_contents);
+  common::FileUtil::createFileWithContent(formatted_checkpoint_local_path, backupDesc, backup_desc_contents);
+  checkpoint_files.emplace_back(backupDesc);
```

Comment:
There is no reason to use emplace_back here, just use push_back

Reply:
Does this guarantee that backup descriptor will only be uploaded after all files get uploaded?